Jint/Runtime/Interop/DefaultTypeConverter.cs

```diff
@@ -18,6 +18,11 @@ public class DefaultTypeConverter : ITypeConverter
         private static readonly ConcurrentDictionary<TypeConversionKey, bool> _knownConversions = new();
         private static readonly ConcurrentDictionary<TypeConversionKey, MethodInfo> _knownCastOperators = new();
 
+        // cache of delegates built against function functions, cannot be static because engine specific
+        // this way we get same delegate instance for each func being used
+        private readonly record struct DelegateCacheKey(Func<JsValue, JsValue[], JsValue> Call, Type DelegateType);
+        private readonly ConcurrentDictionary<DelegateCacheKey,Delegate> _delegateCompilations = new();
```

Comment:
I had a quick look at this and I don't know the answer. But wouldn't this prevent GC? Maybe `ConditionalWeakTable` is more suitable.

Reply:
That's a very good point. I've changed the code to cache the delegate inside the function instance as hidden property. A rare use case so not introducing actual CLR field for this. Now the target and generated delegate should share the same lifetime.

Reply:
The new changes will not work with multiple target types. I think you should revert to the old code and use `ConditionalWeakTable` (which is basically WeakMap) instead of `ConcurrentDictionary`.

I think creating a JS property to keep a C# data feels hacky. It could be a field in `FunctionInstance`. It must be `Dictionary<Type, Delegate>` then. But `ConditionalWeakTable` feels better for this. According to C# docs: 

> The [ConditionalWeakTable<TKey,TValue>](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-6.0) class enables language compilers to attach arbitrary properties to managed objects at run time.

This is a weird and confusing way to say it is a WeakMap. But summarizes the current use case.

Reply:
I think adding the type to key should be enough to work around the uniqueness problem. ConditionalWeaktable probably has more overhead as functions can be short-lived (during single script) whereas the weak table would have engine's lifetime and then the liveness-maintaining.

Reply:
I made the get and set extension method to make it easier to change later. The `ConditionalWeakTable` is a solid choice but I think I'm leaning towards more instance-lifetime for now.  Fixed the key to contain Type's GUID.